harvardsettings_pi.php

```diff
@@ -299,8 +313,79 @@ public function import_users($arr) {
         return array(
             'message' => $message,
             'users_added' => $users_added,
-            'new_accounts_created' => $new_accounts_created
+            'new_accounts_created' => $new_accounts_created,
+            'unsuccessful' => $unsuccessful
+        );
+    }
+    
+    public function get_scalar_user(&$email_row, $huid_row, &$name, $unkown_emails) {
+        if ($email_row === '') {
+            if (!$unkown_emails || !isset($unkown_emails->$huid_row)) {
+                return "no identifiers";
+            }
+            $email_row = $unkown_emails->$huid_row->email;
+            $name = $name === '' ? $unkown_emails->$huid_row->name : $name;   
+        }
+        return $this->CI->users->get_by_email($email_row);
+    }
+    
+    public function new_scalar_user($email_row, $name, &$new_accounts_created) {
+        $new_user = array(
+            'email' => $email_row,
+            'fullname' => $name === '' ? 'placeholder' : $name,
+            'password' => 'DISABLEDPASSWORD',
+        );
+        if (!$this->CI->users->db->insert($this->CI->users->users_table, $new_user)) {
+            return false;
+        }
+        $user = $this->CI->users->get_by_email($email_row);
+        if ($user) {$new_accounts_created++;}
+        return $user;
+    }
+    
+    public function get_pds_emails($url) {
+        $c = curl_init();
+        curl_setopt($c, CURLOPT_URL, $this->CI->config->item('pds_base_url').$url);
+        curl_setopt($c, CURLOPT_RETURNTRANSFER, true);
+        curl_setopt(
+            $c,
+            CURLOPT_USERPWD,
+            $this->CI->config->item('pds_client_id') . ":" . $this->CI->config->item('pds_client_secret')
+        );
+        $output = json_decode(curl_exec($c));
+        curl_close($c);
+        
+        $return_obj = (object) array();
```

Comment:
Since the return object holds a mapping of IDs to names/emails, I'm wondering if it might be better to just use a standard array at the top level. The semantics might be a bit clearer, even if the result is the same (table lookup vs property lookup):

```php
$unkown_emails->$huid_row->name
$unkown_emails[$huid_row]->name
```

Reply:
This makes sense. I'm getting used to the idea of associative arrays, and how they differ from objects.